feat: add systemd provider

Tarow opened the pull request:
Addresses https://github.com/sablierapp/sablier/issues/148

As of now it's still WIP as tests will have to be added etc.

Tarow committed:
feat: add systemd provider

Tarow committed:
fix: read custom properties from unit file

acouvreur commented:
Hey @Tarow !

Thanks for this nice contribution!

Do you think we could add tests using testcontainers ? That should work inside a clean ubuntu image right ? We could write some helper functions to manipulate the systemd files.

I will take a closer look to the files when I have time, I'm quite busy at the moment.

Tarow commented:
Getting systemd to work in Docker containers can be a bit tricky. AFAIK it only works in rootful containers and i'm not sure if/how good it will work with Testcontainers.
I can explore it a bit when i find time, but in case it isn't possible, writing a mock for the systemd interface might be the only option.

Tarow committed:
Merge branch 'main' into main

Tarow committed:
Merge branch 'main' of https://github.com/sablierapp/sablier into main

Tarow committed:
feat: port systemd provider to the new provider interfaces

Tarow commented:
I've updated the provider to the latest interface contract, made a couple improvements and added tests.

Some background: 

Systemd will not report custom properties in unit files (e.g. `X-Sablier` section in this case) when using `systemctl show <unit>` and equivalent library functions.
So in order to get the custom sablier labels, the unit file itself has to be parsed.
The full flow is:
List units -> Get `FragmentPath` for each unit -> parse unit file -> extract sablier labels

To avoid excessive I/O and not parse hundrets of unit files every couple seconds, i added two additional mechanisms for the systemd provider:
1. A new option `provider.systemd.unit-patterns`: The provider will only look at unit files matching this pattern. For example, if only podman quadlets should be managed by sablier, it'd make sense to set this to `podman-*.service`.
Each inspected unit still needs appropriate sablier labels in a [X-Sablier] section (Enable, Group, ...)
2. Cache the sablier labels for each fragment path based on mtime/size. If those didn't change, the unit file is not parsed again. That only requires an `os.stat` and is a lot cheaper than reading and parsing the entire file every cycle.

Both of those together lower file I/O by a lot, so the provider should be a lot more efficient now.

acouvreur commented:
Hello @Tarow thank you for your contribution, can you add an example under `examples/systemd` ?

acouvreur commented:
Looks amazing @Tarow thanks a lot. I'm still reviewing your work. The example is making it really easy

acouvreur commented:
I realize that the tests are not being executed because we need to use SABLIER_SYSTEMD_INTEGRATION.

Can we find a way to always run those tests ? Or is the setup too complicated / impossible ?

acouvreur commented:
Maybe a dedicated github ci job:

```
  integration-systemd:
    name: Integration (systemd)
    runs-on: ubuntu-latest
    steps:
      # checkout + setup-go
      - run: sudo apt-get install -y dbus   # mock tests t.Skip without dbus-daemon
      - run: go test -tags=nomsgpack ./pkg/provider/systemd/...
        env:
          SABLIER_SYSTEMD_INTEGRATION: "1"
```

Tarow commented:
It is definitely possible to always run those. The setup wasn't super trivial (`init.sh` entrypoint for the testcontainer), but it should work consistently now.

The reason why i added the additional environment variable is, that all other integration tests for docker, podman, k3s have a fully isolated container that doesn't really touch the host.
Now for the systemd integration test the provider configures real cgroup resource limits (cpu quota, memory, I/O, ...) on the managed unit. Applying those settings requires write access to the cgroup hierarchy (can also be seen in the testcontainers instructions: https://hub.docker.com/r/jrei/systemd-ubuntu). Because the test writes to the hosts cgroupfs (and could technically change resource limits), i kept it opt-in. If you're fine with always running those anyways, the env variable and the extra GH action i created now could be removed.

acouvreur commented on .github/workflows/integration-systemd.yml:
```diff
@@ -0,0 +1,24 @@
+  push:
+    branches: [main]
+  pull_request:
+
+permissions:
+  contents: read
+
+jobs:
+  integration-systemd:
+    name: Integration (systemd)
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v7
+      - name: Set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version-file: "go.mod"
+      - name: Run systemd integration tests
+        env:
+          SABLIER_SYSTEMD_INTEGRATION: "1"
```
> Error: The actions actions/checkout@v7 and actions/setup-go@v5 are not allowed in sablierapp/sablier because all actions must be pinned to a full-length commit SHA.

you will need to use the pin with version in comment notation for those

Tarow replied:
My bad, fixed that now

acouvreur reviewed (approved):
LGTM!

Waiting green worfklows, then merging!

acouvreur commented:
There might be some conflicts following https://github.com/sablierapp/sablier/pull/1087. Can you look into it @Tarow ? Sorry for the inconvenience

Tarow commented:
No worries, i'll check it out later.

acouvreur reviewed (approved):
LGTM
